src/Microsoft.OData.Client/DataServiceContext.cs

```diff
@@ -3899,21 +3942,54 @@ private void UpdateObjectInternal(object entity, bool failIfNotUnchanged)
                 return;
             }
 
+            if(dependsOnIds != null)
+            {
+                //In DataServiceContext.UpdateObject, we pass entity Id(s) as dependsOnIds.
+                //We will extract ChangeOrder from the entity descriptor based on the Id.
+                //This is because we pass the ChangeOrder as Content-ID to the batch writer.
+                //We will pass the Content-ID(s) as dependsOnIds to the batch writer.
+                List<string> dependsOnIdsAsChangeOrders, dependsOnChangeSetIds;
+
+                GetDependsOnChangeOrdersAndChangeSetIds(
+                    dependsOnIds,
+                    this.EntityTracker.Entities,
+                    this.Model,
+                    out dependsOnIdsAsChangeOrders,
+                    out dependsOnChangeSetIds);
+                resource.DependsOnIds = dependsOnIdsAsChangeOrders;
+                resource.DependsOnChangeSetIds = dependsOnChangeSetIds.Distinct().ToList();

```

Comment:
To avoid this check maybe change the container to a hashset.

Reply:
Done